Enable full ICMP error reporting on Linux

oerdnj opened the pull request:
The Linux kernel suppresses some ICMP error messages by default for UDP
sockets. This commit sets `IP_RECVERR`/`IPV6_RECVERR` on the socket to enable full
ICMP error reporting, hopefully resulting in faster failover to working UDP servers
(DNS servers would be one of the benefactors).

Here's a little background on the issue: https://gitlab.isc.org/isc-projects/bind9/-/issues/835 as reported to BIND 9.

Linux's implementation does not report what it considers "transient" conditions, which is defined as Destination Host Unreachable, Destination Network Unreachable, Source Route Failed and Message Too Big.

libuv should set `setsockopt(fd, IPPROTO_IP, IP_RECVERR, &(int){ 1 }, sizeof(int));`, so the applications using libuv can react much more quickly to network events.

More links / references:
* libc bug and discussion: https://sourceware.org/bugzilla/show_bug.cgi?id=24047
* libc patch: https://sourceware.org/legacy-ml/libc-alpha/2019-02/msg00602.html
* systemd bug and discussion: https://github.com/systemd/systemd/issues/10345
* systemd patch: https://github.com/systemd/systemd/pull/10425
* dns operations discussion: https://lists.dns-oarc.net/pipermail/dns-operations/2019-January/018271.html
* Linux kernel bug and discussion: https://bugzilla.kernel.org/show_bug.cgi?id=202355
* Linux manpage documentation bug: https://bugzilla.kernel.org/show_bug.cgi?id=202369

oerdnj commented:
I could wrap this into a new option, but it seems like a correct thing to do. If there's a Unreachable error, the application always wants to know this as soon as possible, instead of waiting for a timeout.

santigimeno commented on src/unix/udp.c:
```diff
@@ -507,6 +507,32 @@ static int uv__set_reuse(int fd) {
+  struct sockaddr_in sockfd;
+  unsigned int sockfd_len = sizeof(sockfd);
+  if (getsockname(fd, (struct sockaddr*) &sockfd, &sockfd_len) == -1)
+      return UV__ERR(errno);
```
style nit: fix indentation

oerdnj replied:
Fixed

santigimeno commented on src/unix/udp.c:
```diff
@@ -507,6 +507,32 @@ static int uv__set_reuse(int fd) {
+#if defined(IP_RECVERR) && defined(IPV6_RECVERR)
+  int yes;
+  yes = 1;
+  struct sockaddr_in sockfd;
+  unsigned int sockfd_len = sizeof(sockfd);
```
You could use `uv__sockaddr` so it works for both ipv4 and ipv6.

oerdnj replied:
Both fixed by adding second sa_family_t argument to the function.

santigimeno commented on src/unix/udp.c:
```diff
@@ -533,6 +559,8 @@ int uv__udp_bind(uv_udp_t* handle,
     handle->io_watcher.fd = fd;
   }
 
+  uv__set_recverr(fd);
+
```
From the links posted in the issue I'm not sure if this should only be done for `connected` sockets.

oerdnj commented on src/unix/udp.c:
```diff
@@ -533,6 +559,8 @@ int uv__udp_bind(uv_udp_t* handle,
     handle->io_watcher.fd = fd;
   }
 
+  uv__set_recverr(fd);
```
This was copied verbatim from `uv__set_reuse()`...  should I fix that in one go too?

oerdnj commented on src/unix/udp.c:
```diff
@@ -507,6 +507,32 @@ static int uv__set_reuse(int fd) {
+ */
+
+static int uv__set_recverr(int fd) {
+#if defined(IP_RECVERR) && defined(IPV6_RECVERR)
```
Added the opt-in and changed the #ifdef

oerdnj commented:
So, I made the change to be opt-in, fixed the formatting and the code is using already know `sa_family_t`.  I did not fix the `uv__set_reuse()` as it didn't seem appropriate to mix it in, but I can submit a separate PR for that if you want.

santigimeno commented on include/uv.h:
```diff
@@ -613,7 +613,13 @@ enum uv_udp_flags {
+   * Linux. This stops the Linux kernel from supressing some ICMP error messages
+   * and enables full ICMP error reporting for faster failover.
+   */
+  UV_UDP_RECVERR = 16,
```
This flag should also be added to the [documentation](https://github.com/libuv/libuv/blob/v1.x/docs/src/udp.rst)

oerdnj replied:
The flag has been added to documentation.

santigimeno commented on src/unix/udp.c:
```diff
@@ -507,6 +507,29 @@ static int uv__set_reuse(int fd) {
+ * error reporting, hopefully resulting in faster failover to working name
+ * servers.
+ */
+
```
Syle nit: remove empty line

santigimeno commented on test/test-udp-send-unreachable.c:
```diff
@@ -59,12 +59,11 @@ static void close_cb(uv_handle_t* handle) {
 static void send_cb(uv_udp_send_t* req, int status) {
   ASSERT(req != NULL);
-  ASSERT(status == 0);
+  ASSERT(status == 0 || status == UV_ECONNREFUSED);
```
We could expand this test with a new client that uses the new flag.

oerdnj replied:
I expanded the test, but the condition needs to stay, as there's no guarantee that the kernel will actually return `UV_ECONNREFUSED` when running the test.

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1936/

mnordhoff commented:
D:

bnoordhuis commented on src/unix/udp.c:
```diff
@@ -507,6 +507,28 @@ static int uv__set_reuse(int fd) {
+    if (setsockopt(fd, IPPROTO_IPV6, IPV6_RECVERR, &yes, sizeof(yes)))
+       return UV__ERR(errno);
+  }
+#endif
+  return 0;
```
```suggestion
  return 0;
#endif
  return UV_ENOSYS;
```
Unless there's a compelling reason to keep it a no-op on non-linux.

oerdnj replied:
That change would require every client application to apply the `UV_UDP_RECVERR` selectively, so any client application using `UV_UDP_RECVERR` would have to do:
```
#ifdef __linux__
uv_bind(..., UV_UDP_RECVERR);
#else
uv_bind(..., 0);
#endif
```

I don't think this is desirable.

bnoordhuis replied:
Returning `UV_ENOSYS` for unsupported features is libuv policy but let's see what @libuv/collaborators think.

santigimeno replied:
While it's true that's the usual policy, there are some exceptions such in some [process flags](http://docs.libuv.org/en/v1.x/process.html#c.uv_process_flags) that are windows-only and silently ignored in other platforms. Maybe renaming it to `UV_UDP_LINUX_RECVERR` can be an acceptable solution.

bnoordhuis replied:
Good point and good middle ground. What do you think, @oerdnj?

oerdnj replied:
Ok, fair enough.

oerdnj replied:
Renamed in e714fe79

bnoordhuis commented on test/test-udp-send-unreachable.c:
```diff
@@ -138,12 +161,54 @@ TEST_IMPL(udp_send_unreachable) {
+  ASSERT(close_cb_called == 3);
+#else
   ASSERT(close_cb_called == 2);
+#endif
```
Not a blocker but I'd prefer to get rid of the #ifdefs and use a flag. One or two #ifdefs aren't a problem but a lot are. What I mean is this:
```c
TEST_IMPL(udp_send_unreachable) {
  // ...
  int can_recverr;
  // ...
#ifdef __linux__
  can_recverr = 1;
#else
  can_recverr = 0;
#endif
  // ...
  if (can_recverr) {
    // ...
  }
  // ...
  ASSERT_EQ(send_cb_called, can_recverr ? 4 : 2);
  // ...
}
```
Feel free to update any ASSERT statements you touch to the newer ASSERT_EQ macro.

oerdnj replied:
@bnoordhuis Ok, I did that in a fixup commit, let me know if it's what you had on mind.

oerdnj replied:
Isn't this the case that it's actually no-op on the other platforms because the ICMP reporting is already enabled there and doesn't require a special flag?

At least the `send()` on FreeBSD already returns [ECONNREFUSED](https://www.freebsd.org/cgi/man.cgi?query=send&sektion=2&apropos=0&manpath=FreeBSD+12.1-RELEASE+and+Ports) as does `send()` on [OpenBSD](https://man.openbsd.org/sendmsg.2).

bnoordhuis commented on docs/src/udp.rst:
```diff
@@ -47,6 +47,13 @@ Data types
+            /*
+             * Indicates if IP_RECVERR/IPV6_RECVERR will be set when binding the handle.
+             * This sets IP_RECVERR for IPv4 and IPV6_RECVERR for IPv6 UDP sockets on
+             * Linux. This stops the Linux kernel from supressing some ICMP error messages
```
```suggestion
             * Linux. This stops the Linux kernel from suppressing some ICMP error messages
```

oerdnj replied:
Fixed up in f765a9df

bnoordhuis commented on include/uv.h:
```diff
@@ -613,7 +613,13 @@ enum uv_udp_flags {
+  /*
+   * Indicates if IP_RECVERR/IPV6_RECVERR will be set when binding the handle.
+   * This sets IP_RECVERR for IPv4 and IPV6_RECVERR for IPv6 UDP sockets on
+   * Linux. This stops the Linux kernel from supressing some ICMP error messages
```
```suggestion
   * Linux. This stops the Linux kernel from suppressing some ICMP error messages
```

oerdnj replied:
Fixed up in f765a9df

bnoordhuis commented on test/test-udp-send-unreachable.c:
```diff
@@ -138,12 +155,44 @@ TEST_IMPL(udp_send_unreachable) {
-  ASSERT(recv_cb_called == alloc_cb_called);
-  ASSERT(timer_cb_called == 1);
-  ASSERT(close_cb_called == 2);
+  ASSERT_EQ(send_cb_called, (long)(can_recverr ? 4 : 2));
```
Is the cast necessary? And if so, why `long` instead of `int`? `send_cb_called` is of type int.

oerdnj replied:
Changed the cast to `(int)` in 99ca2862049961e0156f2fbd359074842aa8152b, but I actually think there was a warning compiling the code, otherwise I wouldn't add the cast :).

mnordhoff commented:
D:

mnordhoff commented:
[random emoticon]

oerdnj commented:
Ok, I think I addressed all the comments now.  I am going to rebase and squash the branch.

santigimeno commented on test/test-udp-send-unreachable.c:
```diff
@@ -138,12 +155,45 @@ TEST_IMPL(udp_send_unreachable) {
+    /* client sends "PING", then "PANG" */
+    buf = uv_buf_init("PING", 4);
+
+    r = uv_udp_send(&req3,
+		    &client2,
+		    &buf,
+		    1,
+		    (const struct sockaddr*) &addr,
+		    send_cb_recverr);
```
style nit: align arguments

santigimeno commented on test/test-udp-send-unreachable.c:
```diff
@@ -138,12 +155,45 @@ TEST_IMPL(udp_send_unreachable) {
                   send_cb);
   ASSERT(r == 0);
 
+  r = uv_udp_init(uv_default_loop(), &client2);
+  ASSERT(r == 0);
+
+  r = uv_udp_bind(&client2, (const struct sockaddr*) &addr3,
+		  UV_UDP_LINUX_RECVERR);
+  ASSERT(r == 0);
+
+  if (can_recverr) {
+    r = uv_udp_recv_start(&client2, alloc_cb, recv_cb);
+    ASSERT(r == 0);
+
```
Nit: it would be great in the new code use the new `ASSERT` macros

oerdnj replied:
Done

santigimeno commented on docs/src/udp.rst:
```diff
@@ -47,6 +47,13 @@ Data types
              * must not be freed by the recv_cb callback.
              */
             UV_UDP_MMSG_CHUNK = 8,
+            /*
+             * Indicates if IP_RECVERR/IPV6_RECVERR will be set when binding the handle.
+             * This sets IP_RECVERR for IPv4 and IPV6_RECVERR for IPv6 UDP sockets on
+             * Linux. This stops the Linux kernel from suppressing some ICMP error messages
+             * and enables full ICMP error reporting for faster failover.
+             */
+            UV_UDP_LINUX_RECVERR = 16,
```
I understand it's implied but may make clear that it's a noop in other platforms.

oerdnj replied:
Ok, added a short note.

santigimeno reviewed (approved):
LGTM with some nits. Thanks!

santigimeno commented on test/test-udp-send-unreachable.c:
```diff
@@ -138,12 +155,45 @@ TEST_IMPL(udp_send_unreachable) {
+  r = uv_udp_init(uv_default_loop(), &client2);
+  ASSERT(r == 0);
+
+  r = uv_udp_bind(&client2, (const struct sockaddr*) &addr3,
+		  UV_UDP_LINUX_RECVERR);
```
Also align arguments here (usually 1 per line)

oerdnj replied:
Ok.

Just a suggestion, I am not arguing here - what we did in BIND 9 was that we wrote close enough clang-format rules and went with that. Now instead of having formatting as part of the review, we have automated rule that fails CI if somebody forgets to reformat the new code (there are hooks for most of the common editors and git-clang-format tool to help with workflow). This is something that you might want to consider as a group because it is very helpful.

oerdnj replied:
Ok, added a note there, is that enough?

oerdnj replied:
Hopefully, now everything is aligned.

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2039/

cjihrig commented:
Just a heads up, the last CI run seemed to fail a lot across the board. There are also merge conflicts now.

oerdnj commented:
I rebased and squashed all the fixup commits. The tests pass on my local linux system.

bnoordhuis commented:
Still LGTM and thanks for rebasing, @oerdnj. I think this is good to go if the CI doesn't turn up anything untoward.

CI: https://ci.nodejs.org/job/libuv-test-commit/2082/

santigimeno commented:
@oerdnj I think something like [this](https://github.com/santigimeno/libuv/commit/9124b8da42ae5cbef1964cef40656f7989896ebe) should fix the test failures.

oerdnj commented:
> @oerdnj I think something like [this](https://github.com/santigimeno/libuv/commit/9124b8da42ae5cbef1964cef40656f7989896ebe) should fix the test failures.

Thanks, I cherry-picked this to the branch and I will squash it after the CI finishes.

oerdnj commented:
I also have a general question - there are couple other useful socket options for writing high performance servers:

* `IP_FREEBIND` - allows the socket to be bound even if the address doesn't exist (yet) - resolves issues with IPv6 DAD
* `SO_REUSEPORT_LB` - on FreeBSD 12 (and matching SO_REUSEPORT on Linux)
* `SO_INCOMING_CPU` - sets socket affinity to CPU
* `TCP_USER_TIMEOUT` - on Linux this allows setting the TCP connection timeout (but also interacts with KEEPALIVE) - there's different option for TCP connection timeout on Windows, different option (well two in fact) on macOS and different option on FreeBSD (no option on OpenBSD and NetBSD) - while it's certainly possible to have `uv_timer_t` cancel the `uv_tcp_connect()`, the faster timeout in kernel is also useful.
* `TCP_KEEPALIVE` options - this is quite complicated matter, but at least Linux and FreeBSD is keeping up and it might be worthwhile to put some effort into adding high level interface to it

And my question is - is it useful for you and me to put an effort into adding these into libuv?  All the options can be used as of now with:

```
fd = socket(...);
setsockopt(fd, ...);
uv_<proto>_init(..., handle);
uv_<proto>_open(handle, fd);
```

But I am fighting the entropy (code duplication) where I can and it's not very useful if all the project using libuv will have to do this in the end.

santigimeno commented:
> I also have a general question - there are couple other useful socket options for writing high performance servers:
>

 I would say that as long as the options are supported in, at least, a couple of tier 1 platforms it should be fine (that's only my opinion though).

> `TCP_KEEPALIVE` options - this is quite complicated matter, but at least Linux and FreeBSD is keeping up and it might be worthwhile to put some effort into adding high level interface to it

Does http://docs.libuv.org/en/v1.x/tcp.html#c.uv_tcp_keepalive work for you?

santigimeno commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2096/

santigimeno commented:
Another CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2128/

santigimeno commented:
CI looks good. I'd like to have this merged for the next release, @oerdnj could you please rebase and squash the commits. Thanks!

oerdnj commented:
> CI looks good. I'd like to have this merged for the next release, @oerdnj could you please rebase and squash the commits. Thanks!

Done, rebased and squashed. Thanks.

santigimeno commented:
libuv-in-node CI: https://ci.nodejs.org/view/libuv/job/libuv-in-node/181/

santigimeno commented:
Landed in c382d39a85c1edbaf96f0cd84385cd6549c0bde2 and 285a5ea819035ff777b8b7c6a367f3f5b55d8809. Thanks!
